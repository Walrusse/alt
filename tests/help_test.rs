#![cfg(test)]

mod test_env;
use clap::crate_version;
use serde::Serialize;
use std::{convert::TryFrom, error::Error, process::Output};
use test_case::test_case;
use test_env::TestEnv;

type TestResult = Result<(), Box<dyn Error>>;

#[derive(Debug, Serialize)]
struct OutputSnapshot {
    status: i32,
    stdout: String,
    stderr: String,
}

impl TryFrom<Output> for OutputSnapshot {
    type Error = Box<dyn Error>;

    fn try_from(value: Output) -> Result<Self, Self::Error> {
        let res = OutputSnapshot {
            status: value.status.code().ok_or("unable to read status code")?,
            stdout: String::from_utf8(value.stdout)?,
            stderr: String::from_utf8(value.stderr)?,
        };

        Ok(res)
    }
}

#[test_case(vec!["help"]; "help command")]
#[test_case(vec!["--help"]; "long help flag")]
#[test_case(vec!["-h"]; "short help flag")]
#[test_case(vec!["def", "--help"]; "def long help flag")]
#[test_case(vec!["def", "-h"]; "def short help flag")]
#[test_case(vec!["doctor", "--help"]; "doctor long help flag")]
#[test_case(vec!["doctor", "-h"]; "doctor short help flag")]
#[test_case(vec!["exec", "--help"]; "exec long help flag")]
#[test_case(vec!["exec", "-h"]; "exec short help flag")]
#[test_case(vec!["scan", "--help"]; "scan long help flag")]
#[test_case(vec!["scan", "-h"]; "scan short help flag")]
#[test_case(vec!["shim", "--help"]; "shim long help flag")]
#[test_case(vec!["shim", "-h"]; "shim short help flag")]
#[test_case(vec!["show", "--help"]; "show long help flag")]
#[test_case(vec!["show", "-h"]; "show short help flag")]
#[test_case(vec!["use", "--help"]; "use long help flag")]
#[test_case(vec!["use", "-h"]; "use short help flag")]
#[test_case(vec!["which", "--help"]; "which long help flag")]
#[test_case(vec!["which", "-h"]; "which short help flag")]
fn test_help(args: Vec<&str>) -> TestResult {
    let env = TestEnv::new();
    let output = env.alt().args(&args).output()?;

    let snapshot = OutputSnapshot::try_from(output)?;
    insta::with_settings!({
        filters => vec![(regex::escape(crate_version!()).as_str(), "[version]")],
        snapshot_suffix => args.join("_"),
    }, {
        insta::assert_toml_snapshot!(snapshot);
    });

    Ok(())
}
