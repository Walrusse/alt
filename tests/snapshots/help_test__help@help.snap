---
source: tests/help_test.rs
expression: snapshot
---
status = 0
stdout = '''
alt [version]
Switch between different versions of commands

USAGE:
    alt <SUBCOMMAND>

OPTIONS:
    -h, --help       Print help information
    -V, --version    Print version information

SUBCOMMANDS:
    def       Define a new version
    doctor    Checks if alt is setup correctly. Helps debug problems.
    exec      Run the given command
    help      Print this message or the help of the given subcommand(s)
    scan      Scan for different versions of the given command
    shim      Generate shims for all managed commands
    show      Print commands and their versions
    use       Switch the version of a command
    which     Print the resolved path of a command
'''
stderr = ''
